diff --git a/CMakeLists.txt b/CMakeLists.txt
index 21607e9..1cdaa07 100644
--- "a/CMakeLists.txt"	
+++ "b/CMakeLists.txt"	
@@ -12,7 +12,7 @@
 
-cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.28)
 file(STRINGS "VERSION.txt" SPM_VERSION)
 message(STATUS "VERSION: ${SPM_VERSION}")
 
@@ -181,7 +181,7 @@ elseif (SPM_ABSL_PROVIDER STREQUAL "package")
 endif()
 
 add_subdirectory(src)
-add_subdirectory(third_party)
+# add_subdirectory(third_party)
 
 set(CPACK_SOURCE_GENERATOR "TXZ")
 set(CPACK_GENERATOR "7Z")
