diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7b11178..7749fa9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -115,10 +115,11 @@ if(TRITON_ENABLE_CC_HTTP OR TRITON_ENABLE_CC_GRPC OR TRITON_ENABLE_PERF_ANALYZER
     file(STRINGS /etc/os-release DISTRO REGEX "^NAME=")
     string(REGEX REPLACE "NAME=\"(.*)\"" "\\1" DISTRO "${DISTRO}")
     message(STATUS "Distro Name: ${DISTRO}")
-    if(DISTRO STREQUAL "CentOS Linux")
+    if(DISTRO STREQUAL "CentOS Linux" AND CMAKE_SIZEOF_VOID_P EQUAL 8)
       set (CURL_LIB_DIR "lib64")
     endif()
   endif()
+  message(STATUS "Triton client CURL_LIB_DIR=${CURL_LIB_DIR}")
 
   set(_cc_client_depends "")
   if(${TRITON_ENABLE_CC_HTTP})
diff --git a/src/c++/library/CMakeLists.txt b/src/c++/library/CMakeLists.txt
index bdaae25..c36dbc8 100644
--- a/src/c++/library/CMakeLists.txt
+++ b/src/c++/library/CMakeLists.txt
@@ -320,7 +320,7 @@ if(TRITON_ENABLE_CC_HTTP OR TRITON_ENABLE_PERF_ANALYZER)
        httpclient
        PROPERTIES
          LINK_DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/libhttpclient.ldscript
-         LINK_FLAGS "-Wl,--version-script=libhttpclient.ldscript"
+         LINK_FLAGS "-Wl,--version-script=${CMAKE_CURRENT_BINARY_DIR}/libhttpclient.ldscript"
      )
   endif() # NOT WIN32
 
